### PR TITLE
Match cblas behaviour in case of errors with rotmg

### PR DESCRIPTION
**Pull request by Rbiessy**

### CODE REVIEW

**Review by fabiomestre — COMMENTED** [review]
I'm not a big fan of changing the implementation to what (I think) is a worse one just to please clblas. This makes it impossible to distinguish an error value from a valid result. In my opinion, the oneMKL test should be changed instead if possible. 

The algorithm used by cblas is also different. So, if we are choosing this approach, there is more places where clblas behaviour differs from sycl-blas. This are the ones that I found during implementation:

Valid inputs:
- cblas does not seem to be rescaling d1 and d2 in the (0, c) scenario
- I believe clblas is based on netlib implementation,  which means that some specific inputs will give wrong results - https://github.com/xianyi/OpenBLAS/issues/1452

Invalid inputs such as overflows:
- There is potential for an infinite-loop when overflows happen
-  If d1 is less than 0 then the results are implementation defined. Unfortunately, cblas is not consistent with the outputs

**Comment by fabiomestre** [comment]
> Thanks Fabio, I am not a fan either. I considered changing the rotmg tests in oneMKL but I think it is important to match cblas as often as possible. Is flag=2 a value we introduced ourselves in case of error? It seems cblas never outputs flag=2. I think it is difficult to argue that we want to make the oneMKL test less restrictive because we introduce a new value for the flag in SYCL-BLAS.
> 
> I am not sure if the issue in the (0, c) scenario is a bug or something we are missing but I think this case is unlikely to happen as the inputs are random in oneMKL. The invalid inputs cases are avoided in oneMKL tests so they're not an issue for us. The bug you mention is worrying. I'm not sure what we can do about it, maybe we would need to test some specific values in oneMKL like we do in SYCL-BLAS.

I think matching clblas makes sense in most situations. However, for rotmg I would personally make an exception. The rotmg specification is very vague (it doesn't specificy any error code) and I believe most implementations just "copy and paste" the sample netlib code (including clblas).

The problem with that is that the netlib code is very hard to understand and, presumably, wrong in places (as discussed in that OpenBLAS github link). To fix those issues, they ended up rewriting their implementation to match the one found in this paper: https://kar.kent.ac.uk/21699/1/restructuring_the_blas_level_hopkins.pdf

This was the approach that I used when writing rotmg for SYCL-BLAS. I thought that if OpenBLAS had to change their implementation, might as well do the same by default on SYCL-BLAS. So the error code = 2 comes directly from that paper.

What I'm trying to argue is that the only reason why we only have to modify the error code (instead of the whole algorithm) to match clblas is because the test in oneMKL is very basic and always uses the same seed. If the tests were more extensive or used a different seed, then there would be many more changes needed.

That said, I don't necessarily oppose this solution if changing the test is not viable. Just trying to warn that matching clblas might not be the best way of handling the rotmg implementation.

**Comment by Rbiessy** [comment]
Ok it's good know that we're not the only implementation with this behaviour then.

> What I'm trying to argue is that the only reason why we only have to modify the error code (instead of the whole algorithm) to match clblas is because the test in oneMKL is very basic and always uses the same seed. If the tests were more extensive or used a different seed, then there would be many more changes needed.

It uses the same seed but the order in which you run the tests change the internal state of the random generator which ends up producing different values. In other words these tests were passing for me when I ran only the rotmg tests but running all of them produced different inputs and ended up failing.

In that case how do you suggest we change the oneMKL tests? Remove the check that the [flag match](https://github.com/oneapi-src/oneMKL/pull/264/files#diff-351ffde97f9feb382580dd4a15111153c7eea25ea2a9af9396c1baaf7ead195cR158)? Fix the input values?
It may be easier to merge this for now and revert the change after we manage to update the oneMKL tests.

File: src/operations/blas1_trees.hpp [hunk 1]
```diff
@@ -850,7 +850,11 @@ SYCL_BLAS_INLINE typename Rotmg<operand_t>::value_t Rotmg<operand_t>::eval(
       h22_ref = h22;
     }
   }
-  flag_ref = flag;
+  // cblas sets the flag to -1 in case of errors
```
[the remark this answers is not in the document]
Comment (Rbiessy) [reply src/operations/blas1_trees.hpp:853]: Yes [cblas](https://github.com/codeplaysoftware/sycl-blas/blob/master/external/cblas/include/cblas.h) is netlib blas which we use for our tests. oneMKL also uses cblas for the tests and it has the same behaviour for the cases we've tested so far at least.
I don't know about cublas, I'll see if I can test that easily.

**Comment by fabiomestre** [comment]
> r in which you run the tests change the internal state of the random generator which ends up producing different values. In other words these tests were passing for me when I ran only the rotmg tests but running all of them produced different inputs and ended up failing.
> 
> In that case how do you suggest we change the oneMKL tests? Remove the check that the [flag match](https://github.com/oneapi-src/oneMKL/pull/264/files#diff-351ffde97f9feb382580dd4a15111153c7eea25ea2a9af9396c1baaf7ead195cR158)? Fix the input values? It may be easier to merge this for now and revert the change after we manage to update the oneMKL tests.

I think merging and reverting later is ok. A possible "fix" to the tests could be:

1 - Run the output of rotmg on clblas as input to rotm
2 - If rotm does not 0 the y coordinate, then it's an error / overflow. Ignore checks since they are implementation dependent?
3 - Otherwise, run the output of oneMKL implementation as input to rotm. It has to 0 the y coordinate to be valid.

I believe this should cover most cases except the cases where clblas is buggy.

File: src/operations/blas1_trees.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Rbiessy) [reply src/operations/blas1_trees.hpp:853]: cublas is also returning -1 in this case, like cblas.

**Comment by Rbiessy** [comment]
@fabiomestre I may be missing something but it sounds like with this plan we won't be able to compare the flag value which is still an issue. I'll merge now to match the behaviour with most implementations.

File: src/operations/blas1_trees.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mehdi-goli) [reply src/operations/blas1_trees.hpp:853]: Thanks for checking so we have to do the same

**Pull request merged**